File: Week_0/[solution]/크레인 인형뽑기 게임.py
"""
author : Park Min Hyeok
github : https://github.com/m1nnh
e-mail : alsgur9784@naver.com

title : 크레인 인형뽑기 게임
description : 구현
"""

from typing import Final


def solution(board, moves):
    INIT_NUMBER: Final = 0
    CHECK_NUMBER: Final = 0
    ADD_NUMBER: Final = 2

    answer = INIT_NUMBER
    stack = []

    for i in moves:
        result = 0

        for j in range(len(board)):
            if board[j][i - 1] != CHECK_NUMBER:
                result = board[j][i - 1]
                board[j][i - 1] = INIT_NUMBER
                break

        if result == CHECK_NUMBER:
            continue
        if not stack:
            stack.append(result)
        else:
            last = stack.pop()

            if last == result:
                answer += ADD_NUMBER
            else:
                stack.append(last)
                stack.append(result)

    return answer


if __name__ == "__main__":
    print(solution([[0, 0, 0, 0, 0], [0, 0, 1, 0, 3], [0, 2, 5, 0, 1], [4, 2, 4, 4, 2], [3, 5, 1, 3, 1]],
                   [1, 5, 3, 5, 1, 2, 1, 4]))

File: Week_0/SeJun/크레인 인형뽑기 게임.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    stack<int> st;
    int cnt = 0;
    int answer = 0;


    while (cnt != moves.size()) {
        int index = moves[cnt] - 1;
        int target;
        for (int i = 0; i < board.size(); i++) {
            if (board[i][index] != 0) {
                target = board[i][index];
                if (!st.empty()) {
                    if (st.top() == target) {
                        answer += 2;
                        st.pop();
                    }
                    else {
                        st.push(target);
                    }
                }
                else {
                    st.push(target);
                }
                board[i][index] = 0;
                break;
            }
        }
        cnt++;
    }

    return answer;
}